### PR TITLE
mastodon: 3.5.3 -> 4.0.2

### PR DESCRIPTION
**Pull request by erictapen**
###### Description of changes

Let's put up some code, so people can test the new release before it is merged.

See the [4.0.0rc1 Changelog](https://github.com/mastodon/mastodon/releases/tag/v4.0.0rc1) for all changes.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [x] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Izorkin** [comment]
I tested revision 1253227. I think there should be no problems with rc2.

**Review by Izorkin — APPROVED** [review]

**Comment by erictapen** [comment]
The integration test now runs through, so this might go smoothly.

**Review by Izorkin — APPROVED** [review]

**Comment by Izorkin** [comment]
Updated to 4.0.1.

**Comment by Izorkin** [comment]
We need to add a variable `SKIP_POST_DEPLOYMENT_MIGRATIONS=true` to `mastodon-init-db` service?

**Comment by piperswe** [comment]
> We need to add a variable `SKIP_POST_DEPLOYMENT_MIGRATIONS=true` to `mastodon-init-db` service?

I think that's only required for zero-downtime deploys (you do the pre-deployment migrations with that environment variable, then unset it and do the post-deployment migrations after deploying the new code), but since it looks like NixOS doesn't do a zero-downtime deployment (instead it shuts down the service, runs the initialization services incl. migration, then starts the service once that's finished) we can probably just do a regular `rails db:migrate`.

File: pkgs/servers/mastodon/version.nix [hunk 1]
```diff
@@ -1 +1 @@
-"3.5.3"
+"4.0.0rc4"
```
Comment (turion) [inline pkgs/servers/mastodon/version.nix:1]: We can bump to 4.0.2 already :sweat_smile:

**Comment by turion** [comment]
Would it be too much work to have `mastodon_3` on version 3.5.5 in parallel? So people could downgrade or delay upgrade easily?

**Comment by turion** [comment]
It seems that only the integration test aarch64-linux fails because of dependency `step-ca-0.23.0`. That's weird, because it successfully built on hydra: https://hydra.nixos.org/build/198762711#tabs-details Let's try again.

@ofborg build mastodon

**Comment by turion** [comment]
@ofborg build mastodon.passthru.tests

**Comment by turion** [comment]
Ok, tests work on CI as well now :shrug: looking great then :)

**Comment by erictapen** [comment]
> Would it be too much work to have `mastodon_3` on version 3.5.5 in parallel? So people could downgrade or delay upgrade easily?

I'd rather not maintain a second release, as I don't see the benefit. The way I see it nobody wants to run the old version for a long time, but only may want to delay a release a bit? This is much more easiliy achieve by using a Nixpkgs pin (e.g. as a flake input) and then overriding mastodon in an overlay.

**Comment by erictapen** [comment]
@Izorkin Could it be that this upgrade is dependent on https://github.com/NixOS/nixpkgs/pull/198130?

**Comment by turion** [comment]
To add some eyeballs, I'll list a few things here that I believe are relevant from the upstream changelog. Hope that's fine :)

### Relevant points from the update notes from https://github.com/mastodon/mastodon/releases/tag/v4.0.0:

* [x] Ruby 3.0.4
* [ ] Does sidekiq automatically have an `ingress` queue after upgrade?

#### Possibly to be added to the NixOS changelog

* [ ] A database backup is recommended. Maybe there is a simple command we could quote. Something like `pg_dump -Fc -U postgres postgres > name_of_the_backup.dump` (taken from update notes)
* [ ] If users customize sidekiq queues, they need to make sure they have an `ingress` queue.

## Sundry

* [ ] Feel free to add me as a maintainer. (I'm still learning about maintaining mastodon as you can see though)

**Comment by turion** [comment]
> Could it be that this upgrade is dependent on #198130?

At least it is intended by upstream :/ the PR was merged before 4.0.0 was released. But the integration test works, so the old nginx config can't be completely broken. Maybe we can slightly extend the integration test to see whether it still works well, and work on that PR independently?

**Comment by Izorkin** [comment]
> @Izorkin Could it be that this upgrade is dependent on #198130?

Can be used independently.

**Review by Izorkin — APPROVED** [review]

**Review by delroth — APPROVED** [review]
Updated from 3.5.3 -> 4.0.2 by applying this PR locally, everything seems to have worked.

> Does sidekiq automatically have an ingress queue after upgrade?

It does.
image

**Review by piperswe — APPROVED** [review]

File: nixos/doc/manual/release-notes/rl-2211.section.md [hunk 2]
```diff
@@ -379,6 +379,8 @@ Available as [services.patroni](options.html#opt-services.patroni.enable).
 
 - There is a new module for the `xfconf` program (the Xfce configuration storage system), which has a dbus service.
 
+- The Mastodon package got upgraded from the major version 3 to 4. See the [v4.0.0 release notes](https://github.com/mastodon/mastodon/releases/tag/v4.0.0) on migration recommendations.
+
```
Comment (Izorkin) [inline nixos/doc/manual/release-notes/rl-2211.section.md:383]: In our case, the migration is done automatically.

**Comment by erictapen** [comment]
> #### Possibly to be added to the NixOS changelog
> *

@turion I think I'd rather refrain from providing specific instructions in the release notes, especially if they are not necessary in every case. What do you think about just providing a link to the v4.0.0 release notes, as I just amended?

File: nixos/doc/manual/release-notes/rl-2211.section.md (hunk 2 above)
Comment (erictapen) [reply nixos/doc/manual/release-notes/rl-2211.section.md:383]: Ah right, we should add something to not let users have the expectation that they *have* to run anything by themself.

Comment (erictapen) [reply nixos/doc/manual/release-notes/rl-2211.section.md:383]: Better now?

Comment (Izorkin) [reply nixos/doc/manual/release-notes/rl-2211.section.md:383]: If `services.mastodon.automaticMigrations = true`, then migration is performed automatically.

Comment (turion) [reply nixos/doc/manual/release-notes/rl-2211.section.md:383]: You could add that the user might want to backup the db.

**Comment by erictapen** [comment]
I want to add, that this PR is security relevant, the same way https://github.com/NixOS/nixpkgs/pull/201309 was.


Unfortunately I got covid and I don't feel like being able to continue this PR rn. From my pov this looks mergeable besides the release notes wording that is still being discussed. Feel free to force-push to this branch and merge this if you see fit, everyone!

**Comment by turion** [comment]
Plenty of people have approved this PR. I'll adjust the wording slightly, fix the manual and merge. Thank you very much for the PR @erictapen and get well soon!

**Comment by pingiun** [comment]
Is this version of mastodon bugged? I get the following error:

```
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29] ActionView::Template::Error (undefined local variable or method `role_id' for #<User id: 1, email: "pingiun@frog.fashion", created_at: "2022-02-19 14:58:04.176664000 +0000", updated_at: "2022-11-18 15:58:26.350569000 +0000", admin: true, locale: "en", encrypted_otp_secret: nil, encrypted_otp_secret_iv: nil, encrypted_otp_secret_salt: nil, consumed_timestep: nil, otp_required_for_login: false, last_emailed_at: "2022-04-19 10:00:00.339662000 +0000", otp_backup_codes: nil, account_id: 107825245977676788, disabled: false, moderator: false, invite_id: nil, chosen_languages: nil, created_by_application_id: nil, approved: true, sign_in_token: "d9fJHg", sign_in_token_sent_at: "2022-03-28 12:14:08.316039000 +0000", webauthn_id: nil, sign_up_ip: nil, otp_secret: nil>
Nov 18 18:30:31 bigarm puma[4078109]: Did you mean?  role):
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29]      6:
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29]      7:   %meta{ name: 'applicationServerKey', content: Rails.configuration.x.vapid_public_key }
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29]      8:
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29]      9:   = render_initial_state
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29]     10:   = javascript_pack_tag 'application', crossorigin: 'anonymous'
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29]     11:
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29]     12: .notranslate.app-holder#mastodon{ data: { props: Oj.dump(default_props) } }
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29]
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29] app/models/user.rb:157:in `role'
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29] app/models/account.rb:124:in `user_role'
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29] app/presenters/initial_state_presenter.rb:9:in `role'
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29] app/helpers/application_helper.rb:223:in `render_initial_state'
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29] app/views/shared/_web_app.html.haml:9
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29] app/views/shared/_web_app.html.haml:1
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29] app/views/home/index.html.haml:7
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29] app/controllers/concerns/localized.rb:11:in `set_locale'
Nov 18 18:30:31 bigarm puma[4078109]: [fe3b6a6e-4270-4a96-9525-1c69120fba29] lib/mastodon/rack_middleware.rb:9:in `call'
```

**Comment by Izorkin** [comment]
Run command:
```
sudo systemctl restart mastodon-init-db
```

**Comment by pingiun** [comment]
> Run command:
> 
> ```
> sudo systemctl restart mastodon-init-db
> ```

Thanks so much! I kinda expected this to work automatically via systemd depends or something

**Comment by Izorkin** [comment]
Enabled option?
```
  services.mastodon.automaticMigrations = true;
```



> Thanks so much! I kinda expected this to work automatically via systemd depends or something

This bug fixed in https://github.com/NixOS/nixpkgs/pull/197893
I don't know why your service `mastodon-init-db` didn't start automatically.

**Comment by pingiun** [comment]
> Enabled option?
> 
> ```
>   services.mastodon.automaticMigrations = true;
> ```
> 

Thank you I will enable this option

**Comment by Izorkin** [comment]
> Thank you I will enable this option

This option is enabled by default.

**Pull request merged**